diffrax/_term.py

```diff
@@ -360,6 +395,15 @@ class WeaklyDiagonalControlTerm(_AbstractControlTerm[_VF, _Control]):
         without the "weak". (This stronger property is useful in some SDE solvers.)
     """
 
+    def __init__(self, *args, **kwargs):
+        warnings.warn(
```

Comment:
Sorry to jump in here, but as `typing_extensions` is already a dependancy, can we use the [`@deprecated`](https://typing-extensions.readthedocs.io/en/latest/index.html#typing_extensions.deprecated) decorator on the class instead? This additionally allows your code editor of choice to highlight the deprecation at coding time: 
<img width="417" alt="deprecation_decorator" src="https://github.com/patrick-kidger/diffrax/assets/97948946/ecdd9677-d28d-4c19-85e7-c9accae8aa8f">

To trigger the runtime warning, the `typing_extensions` version in the `pyproject.toml` would need to be bumped to `>=4.9.0` though. What do you think of this @patrick-kidger?

Reply:
Thank you for the heads up -- I think that sounds like an excellent idea! Lets do it. @lockwo?

Reply:
Sg, added

Reply:
I haven't use the deprecated wrapper before, and I am unable to get it to actually surface a warning. I don't see a lot of docs on the matter. Do you have any advice @tttc3 ? That is the only failing test right now (failing since it doesn't raise a warning).

Reply:
I see runtime warnings are not expected here (https://discuss.python.org/t/pep-702-marking-deprecations-using-the-type-system/23036/21?page=2) I will remove the test

Reply:
I would expect the warning to raise as the decorator is a backward compatible port of [`@warnings.deprecate`](https://docs.python.org/3.13/library/warnings.html#warnings.deprecated) from python 3.13.  For example, my original example raises the following warning at runtime:
```bash
demo.py:11: DeprecationWarning: This is deprecated.
  use_deprecated = dont_use()
```

From what I can see, there is some nuanced interaction between the `eqx.Module` initialisation process and the `if cls is arg` check used in `deprecated` to trigger the warning.

Reply:
Ah, `eqx.Module` does indeed do some fairly cunning stuff during the initialisation -- it swaps out the class with a slightly different one that unfreezes it during `__init__` only. This is what makes our modules be frozen most of the time, but still support `self.foo = foo` assignment during `__init__`.

I think in this case, raising a warning is more important than using the `@deprecated` decorator -- so best to put this back as an explicit `warnings.warn` I think!